File: backend/src/InsuraTech.Domain/Claims/Claim.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using InsuraTech.Domain.Common;
using InsuraTech.Domain.Events;
using InsuraTech.Domain.Exceptions;

namespace InsuraTech.Domain.Claims
{
    public sealed class Claim : AggregateRoot
    {
        public const int MaxOpenClaimsPerPolicy = 3;

        public Guid PolicyId { get; private set; }
        public ClaimType Type { get; private set; }
        public ClaimStatus Status { get; private set; }
        public decimal ClaimedAmount { get; private set; }
        public decimal? ApprovedAmount { get; private set; }
        public DateOnly IncidentDate { get; private set; }
        public string Description { get; private set; } = null!;
        public bool HasBeenAppealed { get; private set; }
        public string? RejectionReason { get; private set; }
        public Guid? CreatedByAdvisorId { get; private set; }

        private readonly List<ClaimStatusHistory> _statusHistory = new();
        public IReadOnlyCollection<ClaimStatusHistory> StatusHistory =>
            _statusHistory.AsReadOnly();
        private Claim() { }
        public static Claim Register(
        Guid policyId,
        ClaimType type,
        decimal claimedAmount,
        DateOnly incidentDate,
        string description,
        DateOnly policyStartDate,
        DateOnly policyEndDate,
        decimal availableInsuredAmount,
        int currentOpenClaimsCount,
        string responsibleUser,
        Guid? createdByAdvisorId = null)
        {
            if (currentOpenClaimsCount >= MaxOpenClaimsPerPolicy)
                throw new ClaimLimitExceededException(policyId);

            if(incidentDate < policyStartDate || incidentDate > policyEndDate)
                throw new ClaimOnExpiredPolicyException(policyId, incidentDate);

            if (claimedAmount > availableInsuredAmount)
                throw new BusinessRuleException("AMOUNT_EXCEEDS_COVERAGE",
                    $"Claimed amount ${claimedAmount} exceeds available insured amount ${availableInsuredAmount}.");

            if (string.IsNullOrWhiteSpace(description))
                throw new ArgumentException("Claim description is required.", nameof(description));

            var claim = new Claim
            {
                PolicyId = policyId,
                Type = type,
                ClaimedAmount = claimedAmount,
                IncidentDate = incidentDate,
                Description = description.Trim(),
                Status = ClaimStatus.PendingApproval,
                HasBeenAppealed = false,
                CreatedByAdvisorId = createdByAdvisorId,
            };

            claim.AddStatusHistory(ClaimStatus.PendingApproval, responsibleUser, "Claim pending leader approval.");
            claim.AddDomainEvent(new ClaimRegisteredEvent(
                claim.Id, policyId, string.Empty, claimedAmount));

            return claim;
        }

        /// <summary>Leader approves the claim — moves from PendingApproval into the normal workflow (Registered).</summary>
        public void ApprovePending(string responsibleUser, string? observations = null)
        {
            if (Status != ClaimStatus.PendingApproval)
                throw new InvalidClaimStateException(Status.ToString(), nameof(ApprovePending));

            var previous = Status;
            Status = ClaimStatus.Registered;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.Registered, responsibleUser, observations ?? "Claim approved by leader.");
            AddDomainEvent(new ClaimStatusChangedEvent(Id, PolicyId, previous, Status, responsibleUser, observations));
        }

        /// <summary>Leader rejects the claim — moves from PendingApproval to Rejected.</summary>
        public void RejectPending(string reason, string responsibleUser)
        {
            if (Status != ClaimStatus.PendingApproval)
                throw new InvalidClaimStateException(Status.ToString(), nameof(RejectPending));

            if (string.IsNullOrWhiteSpace(reason))
                throw new ArgumentException("Rejection reason is required.", nameof(reason));

            var previous = Status;
            Status = ClaimStatus.Rejected;
            RejectionReason = reason;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.Rejected, responsibleUser, reason);
            AddDomainEvent(new ClaimStatusChangedEvent(Id, PolicyId, previous, Status, responsibleUser, reason));
        }

        public void StartInvestigation(string responsibleUser, string? observations = null)
        {
            if (Status != ClaimStatus.Registered)
                throw new InvalidClaimStateException(Status.ToString(), nameof(StartInvestigation));

            var previous = Status;
            Status = ClaimStatus.UnderInvestigation;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.UnderInvestigation, responsibleUser, observations);
            AddDomainEvent(new ClaimStatusChangedEvent(Id, PolicyId, previous, Status, responsibleUser, observations));
        }

        public void Approve(decimal approvedAmount, string responsibleUser, string? observations = null)
        {
            if(Status != ClaimStatus.UnderInvestigation && Status != ClaimStatus.Appealed)
                throw new InvalidClaimStateException(Status.ToString(), nameof (Approve));

            if (approvedAmount <= 0 || approvedAmount > ClaimedAmount)
                throw new ArgumentException($"Approved amount must be between $0 and claimed amount ${ClaimedAmount}.",
                nameof(approvedAmount));

            var previous = Status;
            Status = ClaimStatus.Approved;
            ApprovedAmount = approvedAmount;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.Approved, responsibleUser, observations);
            AddDomainEvent(new ClaimStatusChangedEvent(
                Id, PolicyId, previous, Status, responsibleUser, observations));
        }

        public void Appeal(string responsibleUser, string? observations = null)
        {
            if (Status != ClaimStatus.Rejected)
                throw new InvalidClaimStateException(Status.ToString(), nameof(Appeal));

            if (HasBeenAppealed)
                throw new BusinessRuleException("APPEAL_ALREADY_USED",
                    $"Claim '{Id}' has already been appealed once.");

            var previous = Status;
            Status = ClaimStatus.Appealed;
            HasBeenAppealed = true;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.Appealed, responsibleUser, observations);
            AddDomainEvent(new ClaimStatusChangedEvent(
                Id, PolicyId, previous, Status, responsibleUser, observations));

        }

        public void Reject(string reason, string responsibleUser)
        {
            if (Status != ClaimStatus.UnderInvestigation &&
                Status != ClaimStatus.Appealed)
                throw new InvalidClaimStateException(Status.ToString(), nameof(Reject));

            if (string.IsNullOrWhiteSpace(reason))
                throw new ArgumentException("Rejection reason is required.", nameof(reason));

            var previous = Status;
            Status = ClaimStatus.Rejected;
            RejectionReason = reason;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.Rejected, responsibleUser, reason);
            AddDomainEvent(new ClaimStatusChangedEvent(
                Id, PolicyId, previous, Status, responsibleUser, reason));
        }
        public void RegisterPayment(string responsibleUser, string? observations = null)
        {
            if (Status != ClaimStatus.Approved)
                throw new InvalidClaimStateException(Status.ToString(), nameof(RegisterPayment));

            var previous = Status;
            Status = ClaimStatus.Paid;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(ClaimStatus.Paid, responsibleUser, observations);
            AddDomainEvent(new ClaimStatusChangedEvent(
                Id, PolicyId, previous, Status, responsibleUser, observations));
        }

        public bool IsOpen() => Status is
        ClaimStatus.PendingApproval or
        ClaimStatus.Registered or
        ClaimStatus.UnderInvestigation or
        ClaimStatus.Appealed;

        private void AddStatusHistory(ClaimStatus status, string responsibleUser, string? observations) =>
               _statusHistory.Add(ClaimStatusHistory.Create(Id, status, responsibleUser, observations));
    }
}

File: backend_v2/src/main/java/com/insuratech/domain/claims/Claim.java
package com.insuratech.domain.claims;

import com.insuratech.domain.common.AggregateRoot;
import com.insuratech.domain.events.ClaimRegisteredEvent;
import com.insuratech.domain.events.ClaimStatusChangedEvent;
import com.insuratech.domain.exceptions.ClaimAlreadyClosedException;
import com.insuratech.domain.exceptions.InvalidClaimTransitionException;
import lombok.Getter;
import org.springframework.data.mongodb.core.mapping.Document;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

@Getter
@Document(collection = "claims")
public class Claim extends AggregateRoot {

    private String policyId;
    private ClaimType claimType;
    private ClaimStatus status;
    private String description;
    private BigDecimal claimedAmount;
    private BigDecimal approvedAmount;
    private Instant incidentDate;
    private String reportedBy;
    private String reviewedBy;
    private String rejectionReason;

    private final List<ClaimStatusHistory> statusHistory = new ArrayList<>();

    private static final Set<ClaimStatus> CLOSED_STATUSES = Set.of(
        ClaimStatus.PAID, ClaimStatus.REJECTED, ClaimStatus.CLOSED
    );

    protected Claim() { super(); }

    public static Claim register(String policyId, ClaimType claimType, String description,
                                 BigDecimal claimedAmount, Instant incidentDate, String reportedBy) {
        Claim claim = new Claim();
        claim.policyId = policyId;
        claim.claimType = claimType;
        claim.status = ClaimStatus.REGISTERED;
        claim.description = description;
        claim.claimedAmount = claimedAmount;
        claim.incidentDate = incidentDate;
        claim.reportedBy = reportedBy;
        claim.statusHistory.add(new ClaimStatusHistory(ClaimStatus.REGISTERED, reportedBy, "Claim registered"));
        claim.addDomainEvent(new ClaimRegisteredEvent(claim.getId(), policyId, claimType.name()));
        return claim;
    }

    public void startReview(String reviewedBy) {
        assertNotClosed();
        assertValidTransition(ClaimStatus.UNDER_REVIEW);
        ClaimStatus previous = this.status;
        this.status = ClaimStatus.UNDER_REVIEW;
        this.reviewedBy = reviewedBy;
        statusHistory.add(new ClaimStatusHistory(ClaimStatus.UNDER_REVIEW, reviewedBy, "Review started"));
        markAsUpdated();
        addDomainEvent(new ClaimStatusChangedEvent(getId(), previous.name(), status.name()));
    }

    public void approve(BigDecimal approvedAmount, String approvedBy) {
        assertNotClosed();
        assertValidTransition(ClaimStatus.APPROVED);
        ClaimStatus previous = this.status;
        this.status = ClaimStatus.APPROVED;
        this.approvedAmount = approvedAmount;
        statusHistory.add(new ClaimStatusHistory(ClaimStatus.APPROVED, approvedBy, "Claim approved"));
        markAsUpdated();
        addDomainEvent(new ClaimStatusChangedEvent(getId(), previous.name(), status.name()));
    }

    public void reject(String reason, String rejectedBy) {
        assertNotClosed();
        assertValidTransition(ClaimStatus.REJECTED);
        ClaimStatus previous = this.status;
        this.status = ClaimStatus.REJECTED;
        this.rejectionReason = reason;
        statusHistory.add(new ClaimStatusHistory(ClaimStatus.REJECTED, rejectedBy, reason));
        markAsUpdated();
        addDomainEvent(new ClaimStatusChangedEvent(getId(), previous.name(), status.name()));
    }

    public void markAsPaid(String paidBy) {
        assertNotClosed();
        assertValidTransition(ClaimStatus.PAID);
        ClaimStatus previous = this.status;
        this.status = ClaimStatus.PAID;
        statusHistory.add(new ClaimStatusHistory(ClaimStatus.PAID, paidBy, "Payment processed"));
        markAsUpdated();
        addDomainEvent(new ClaimStatusChangedEvent(getId(), previous.name(), status.name()));
    }

    public void close(String closedBy, String notes) {
        assertNotClosed();
        ClaimStatus previous = this.status;
        this.status = ClaimStatus.CLOSED;
        statusHistory.add(new ClaimStatusHistory(ClaimStatus.CLOSED, closedBy, notes));
        markAsUpdated();
        addDomainEvent(new ClaimStatusChangedEvent(getId(), previous.name(), status.name()));
    }

    private void assertNotClosed() {
        if (CLOSED_STATUSES.contains(status)) throw new ClaimAlreadyClosedException();
    }

    private void assertValidTransition(ClaimStatus target) {
        boolean valid = switch (target) {
            case UNDER_REVIEW -> status == ClaimStatus.REGISTERED;
            case APPROVED, REJECTED -> status == ClaimStatus.UNDER_REVIEW;
            case PAID -> status == ClaimStatus.APPROVED;
            default -> false;
        };
        if (!valid) throw new InvalidClaimTransitionException(status.name(), target.name());
    }

    public List<ClaimStatusHistory> getStatusHistory() {
        return Collections.unmodifiableList(statusHistory);
    }
}
